### DISCUSSION TITLE
7주차 Discussion

### DISCUSSION THREAD

**Discussion by chloe-codes1** [slippStudy/java-network-study#13] [General]
## 📢 책을 읽으면서 궁금한 점들을 자유롭게 적고 대답해 주세요

7주차 진행 방식

> 진행자: 주현님, 진욱님
> 

- 13 ~ 14장 읽기
- Discussion 에 질문 or 인상 깊은 점 2개 적어오기

**Comment by jinwookh** [comment]
1. nettty는 ByteBuffer 대신 독자적인 버퍼 클래스를 사용하고 있다.(ByteBuf, ChannelBuffer)
NIO ByteBuffer와 netty ByteBuf를 비교하는 것도 재밌을 것 같다.(아직 안 해봄)
 https://netty.io/3.5/guide/#architecture.5

2. InterruptibleChannel 내용은 Channel 내용과 중복된다.
void close() throws IOException
InterruptibleChannel를 또 만든 이유가 무엇인지 궁금했다.
답변을 보면 가독성 있는 코드를 위해 만들어진 클래스라고 한다.
https://stackoverflow.com/questions/28366555/whats-the-purpose-of-interruptiblechannel-interface-extends-channel-interface-i

3. java로 만들어진 database인 mysql, lucene등은 파일을 다룰 때 java nio, io 둘 중 무엇을 사용할까?

**Reply by jinwookh** [reply]
4. RandomAccessFile.getChannel()에서 보듯, IO 클래스와 NIO 클래스가 서로 호환이 되도록 되어 있다.
RandomAccessFile을 써야할 때는 언제고, FileChannel을 사용할 때는 언제일까?

**Reply by jinwookh** [reply]
5. 파일 락은 프로그램당 락 하나로 운영될 수 있게끔 설계되었다고 한다.
다수의 이용자가 파일을 조작할 수 있는 서버를 만들려면 어떻게 만들어야 할까?

**Comment by yangyunhyeuk** [comment]
### 인상 깊은 점 
1. 많은 타입(double, float…)들은 프로세스인 JVM이 운영체제와 IO 수행을 위해 통신할 때 데이터 모두를 운영체제가 인식할 수 있는 순차적인 바이트 형태로 바꿔서 전달한다. 또한 시스템 메모리의 구성 형태와 달리 자바에서 사용하는 바이트 배열은 시스템 메모리에서 사용하는 순차적인 바이트들의 집합이 아닌 객체 내 바이트들을 저장하고 있는 형태이기 때문에, 커널이 관리하는 시스템 메모리에 직접 저장할 수 있는 형태가 아닌 것이다. 따라서 이런 한계점을 극복하기 위해 NIO에서 다이렉트 버퍼가 도입되었고 바이트버퍼만이 시스템 메모리를 사용할 수 있는 다이렉트 버퍼로 생성할 수 있다.  

2. 현대의 운영체제들은 빠르게 변화하는 시대에 발맞춰 강력한 캐싱과 프리페치 알고리즘으로 디스크의 IO를 빠르게 처리한다. 그렇지만 비블록킹 모드를 이용할 경우 근본적인 처리 루틴이 달라, 이런 운영체제의 효율적인 기능을 이용하는 데에 제한을 받게 된다. 따라서 파일 채널은 블록킹 모드를 사용하는 것이다.

**Comment by LenKIM** [comment]
인상깊었던 부분은 Byte에 대해서 다이렉트버퍼와 논버퍼가 분리해서 설명하는 부분. 

다이렉터버퍼와 논버퍼의 성능차이는 어떻게 확인할 수 있을지?

**Comment by chloe-codes1** [comment]
### 인상깊은 부분

- Buffer 의 utility method를 살펴볼 수 있어 좋았다
- java.nio의 Buffer class는 `synchronized` keyword가 없어 thread safe하지 않다는 점이 당연하면서도 인상 깊었다
  - Buffer를 사용할 때 동기화가 필요한 부분이 있는지 확인하고 신경써야겠다

**Comment by kangbada0728** [comment]
- 해당 책을 공부하면서 계속 궁금했던게 시스템 버퍼를 자바에서는 활용하지 않나 였는데 Buffer 를 알게되며 조금 의문이 풀렸다.
- NIO 가 나왔으니 당연히 IO 대신 NIO 를 써야할 줄 알았는데 내 생각과 달라서 인상깊었다.
